### ISSUE TITLE
Batch import rules from Blocker.

### ISSUE THREAD

**Issue by tommynok** [MuntashirAkon/AppManager#381]
As far as I know, at the moment **AM** can import blocking rules from **Blocker** one file at a time - one ._json_ file - one application.
With a large number of configured applications (150-200), the import from **Blocker** to **AM** can take quite a lot of time.
Is it possible to implement the function of batch import of rules from Blocker? 
This function would be very convenient for moving from one application to another.
(I hope my request is not a duplicate)

**Comment by MuntashirAkon** [comment]
AM can import one or more files using the import options. To do that, when the Files app is opened, click and hold on an item to enter the batch selection mode (and optionally use the 3-dots menu to select all the rules).

**Comment by tommynok** [comment]
> AM can import one or more files using the import options. To do that, when the Files app is opened, click and hold on an item to enter the batch selection mode (and optionally use the 3-dots menu to select all the rules).

Oh, apparently in older versions of androd this does not work. Through the "Files" stock app, I cannot import any of the Blocker rule files at all.
Short and long tap does not work at all.
In the "files" app on android 5-6, I do not see the "select all" option.
In these OS versions, the three dots menu has options for "Hide storage" and "Hide file size".

**Comment by MuntashirAkon** [comment]
Maybe try MiXplorer or Material Files.

**Comment by tommynok** [comment]
> Maybe try MiXplorer or Material Files.

Mixplorer (and Total Commander) allowed importing only one file at a time.
Only Material Files made it possible to select all rule files.
But the process of importing rules (120 files) was accompanied by abnormal device behavior.
Freezes, launcher reboot, black screen, etc.
On the second attempt to import blocking rules, AM crashed.
[2021-04-21_152553_logcat.txt](https://github.com/MuntashirAkon/AppManager/files/6350820/2021-04-21_152553_logcat.txt)
[2021-04-21_145717_logcat.txt](https://github.com/MuntashirAkon/AppManager/files/6350821/2021-04-21_145717_logcat.txt)
[AMlog.txt](https://github.com/MuntashirAkon/AppManager/files/6350822/AMlog.txt)

**Comment by MuntashirAkon** [comment]
> On the second attempt to import blocking rules, AM crashed.

OutOfMemory (OOM) error. Probably your device doesn't have enough memory to import that many rules at a time (though I'm not sure why the files are kept in the memory in first place).

**Comment by tommynok** [comment]
> OutOfMemory (OOM) error. Probably your device doesn't have enough memory to import that many rules at a time (though I'm not sure why the files are kept in the memory in first place).

The situation is confusing. I tried to remove all rules, and I encountered device freezes and AM crashes.
[AMlog.txt](https://github.com/MuntashirAkon/AppManager/files/6352708/AMlog.txt)
In any case, import rules from MAT into Blocker went quickly and without crashes. But this is probably an _incorrect comparison._

**Comment by MuntashirAkon** [comment]
> [AMlog.txt](https://github.com/MuntashirAkon/AppManager/files/6352708/AMlog.txt)

Logs only contain internal Java classes. I can't do anything with them.

> In any case, import rules from MAT into Blocker went quickly and without crashes.

Do they use SAF too?

**Comment by tommynok** [comment]
> Logs only contain internal Java classes. I can't do anything with them.

This was an unexpected crash, so I have provided only log AM.
I need to reproduce this bug, and then I will add the normal logcat.


> Do they use SAF too?

Yes

**Comment by tommynok** [comment]
> Logs only contain internal Java classes. I can't do anything with them


The error with a crash when deleting all the rules did not happen again.
But the freezes are repeated during this operation.
I made a logcat, I hope it will be useful.
[2021-04-22_192219_logcat_rst.txt](https://github.com/MuntashirAkon/AppManager/files/6361531/2021-04-22_192219_logcat_rst.txt)
I also tried to carry out batch import of blocking rules with a small number of files (20 at a time).
Perhaps this option will be a way out if my device cannot handle importing 120 rules at the same time.
But there is one caveat - when importing, I sometimes get AM error _Could not import 2 (3-4) files_ (log below)
Is it possible to get a report from AM which particular file could not be imported?
[2021-04-23_005609_logcat.txt](https://github.com/MuntashirAkon/AppManager/files/6361536/2021-04-23_005609_logcat.txt)

**Comment by MuntashirAkon** [comment]
> I made a logcat, I hope it will be useful.
> [2021-04-22_192219_logcat_rst.txt](https://github.com/MuntashirAkon/AppManager/files/6361531/2021-04-22_192219_logcat_rst.txt)

Yes, it's very useful indeed. The freezing issue is quite clear from the logs: Too many apps registered the package change receivers. When you're deleting all the rules, it enables a lot of components and each enable operation triggers all these receivers to wake up and do their job. Since all of the receivers use the UI thread AKA main thread, random freezes will occur. If I ever add **Autostarts**-like features, consider blocking all these receivers as they're not only a threat to privacy but also reduces phone's performance as well as battery (fortunately, apps can no longer do that since API 26).

This wasn't a problem in old version of AM where only Intent Firewall was used. But we've found out that IFW is not enough, and as a result, we block components using both Intent Firewall and Package Manager. This is different from Blocker or MAT where you can only use one option at a time, or Watt where you can only use Intent Firewall.

> Is it possible to get a report from AM which particular file could not be imported?

Yeah.

**Comment by tommynok** [comment]
> Too many apps registered the package change receivers. When you're deleting all the rules, it enables a lot of components and each enable operation triggers all these receivers to wake up and do their job.

Now I understand why my phone was not just freezing, but also overheating. I already thought that his last days had come.


> > Is it possible to get a report from AM which particular file could not be imported?
> 
> Yeah.

Oh, that will be great. If the import fails, then at least I can manually configure the components.
The import failure log does not provide any useful information?
I mean the reason why the import failed.

**Comment by MuntashirAkon** [comment]
> Now I understand why my phone was not just freezing, but also overheating. I already thought that his last days had come.

Install **Autostarts** and block all the unnecessary receivers, and your phone will be fast again.

> The import failure log does not provide any useful information?

Unfortunately, no. There could be many thing that can go wrong such as the app in question may not be installed ('cause we have to verify app components as Blocker's backup feature has a bug which [wrongly identifies providers as receivers](https://github.com/MuntashirAkon/AppManager/issues/24#issuecomment-648911307)) or some of the components may not exist at all. In any case, I implemented this feature a long time ago. So, I forgot how I've implemented it. I have to look into it to see how it works.
